From 45d9a4a7667a006a8f614016d9cf4f388d3c8e09 Mon Sep 17 00:00:00 2001
From: Dmitry Artamonow <mad_soft@inbox.ru>
Date: Fri, 13 Mar 2009 20:56:03 +0300
Subject: [PATCH 19/28] h3600: add gpio-keys for Power and Action buttons

This is the only buttons on h3100/3600 connected to GPIOs. Other buttons
are connected via resistor matrix to ADC on Atmel microcontroller.

Signed-off-by: Dmitry Artamonow <mad_soft@inbox.ru>
---
 arch/arm/mach-sa1100/h3600.c |   40 ++++++++++++++++++++++++++++++++++++++++
 1 files changed, 40 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-sa1100/h3600.c b/arch/arm/mach-sa1100/h3600.c
index 96da4e1..c7e0458 100644
--- a/arch/arm/mach-sa1100/h3600.c
+++ b/arch/arm/mach-sa1100/h3600.c
@@ -24,6 +24,9 @@
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/partitions.h>
 #include <linux/serial_core.h>
+#include <linux/gpio_keys.h>
+#include <linux/input.h>
+#include <linux/platform_device.h>
 
 #include <mach/hardware.h>
 #include <asm/mach-types.h>
@@ -413,11 +416,48 @@ static struct irda_platform_data h3600_irda_data = {
 	.set_speed	= h3600_irda_set_speed,
 };
 
+/* GPIO keys */
+
+static struct gpio_keys_button h3600_button_table[] = {
+	{
+		.code		= KEY_POWER,
+		.gpio		= H3600_GPIO_NPOWER_BUTTON,
+		.desc		= "Power Button",
+		.active_low	= 1,
+		.type		= EV_KEY,
+		.wakeup		= 1,
+	}, {
+		.code		= KEY_ENTER,
+		.gpio		= H3600_GPIO_ACTION_BUTTON,
+		.active_low	= 1,
+		.desc		= "Action button",
+		.type		= EV_KEY,
+		.wakeup		= 1,	/* FIXME - for testing only */
+	},
+};
+
+static struct gpio_keys_platform_data h3600_keys_data = {
+	.buttons  = h3600_button_table,
+	.nbuttons = ARRAY_SIZE(h3600_button_table),
+};
+
+static struct platform_device h3600_keys = {
+	.name	= "gpio-keys",
+	.id	= -1,
+	.dev	= {
+		.platform_data = &h3600_keys_data,
+	},
+};
+
+static struct platform_device *h3600_devices[] = {
+       &h3600_keys,
+};
 
 static void h3600_mach_init(void)
 {
 	sa11x0_set_flash_data(&h3xxx_flash_data, &h3xxx_flash_resource, 1);
 	sa11x0_set_irda_data(&h3600_irda_data);
+	platform_add_devices(h3600_devices, ARRAY_SIZE(h3600_devices));
 }
 
 MACHINE_START(H3600, "Compaq iPAQ H3600")
-- 
1.6.1.3

